### PR TITLE
add config system 

### PR DESCRIPTION
**Pull request by L1aoXingyu**
# 配置系统

目前的配置系统基于下面4个特点进行构建：

- 生成的配置文件可以序列化和反序列化，这样的好处有两个：1. 用户可以通过 diff config 轻松获得两次训练的配置差异；2. 直接 load config 便能复现别人的结果，能够节省大量沟通的时间成本；
- 配置系统层次清晰，比如模型结构相关的参数、网络训练的超参以及数据读取的参数能够清晰的分开；
- 参数在配置之后是可修改的，因为有的时候需要根据数据集的 size 再动态设置训练的 iters 等，这就需要用户定义好配置之后，参数可以再次被修改；
- 用户可以灵活增加配置参数而不需要侵入式修改 libai 的内部代码，希望用户可以将 libai 作为一个 lib 进行使用，而不是 clone 之后在上面修改；

## 两种配置方式

最终选择基于 LazyConfig 来构建 libai 项目，提供两种风格的配置文件，一种是注册+解析的方式，一种是 LazyCall 的方式。注册+解析是为了兼容之前一些项目的构建习惯，LazyCall 则是另外一种全新的配置系统思路，比注册+解析更灵活。

注册+解析是之前比较常用的配置方式，通过装饰器将可能会用到的对象注册好，随后便可以通过名字去获取对应的对象，在实例化阶段将参数传入。这种方式的一个问题在于额外引入的注册机制并不够方便，只有注册过的对象才能获取，而我们显然不能对所有代码中的对象都进行注册，一些其他库的对象也无法注册；另外一个问题在于不同用户需要使用对方注册的内容时也不够方便。

LazyCall 是另外一种类型的配置方式，不需要利用注册系统，利用 python 配置文件的特性将需要的模块直接 import，利用延后初始化的特性直接传参，在实例化之前可以进行任意参数的修改，最终在需要的时候进行实例化。

## 配置文件语法

上述两种方式都是采用了 python 文件来构建，因为序列化和反序列化的问题，没有采用命令行传参。另外考虑到 yaml 文件不够灵活，而 python 不仅语法更熟悉，而且可以内置一些计算，更复杂的嵌套和数据类型。

最终生成的配置文件会保存成一个 yaml 文件，如果要复现实验结果，只需要将 `config.yaml` 作为配置文件传到系统中。

不同类型的参数可以放在不同的地方实现参数隔离，比如

```python
# data.py
data = dict(
    # Pad the vocab size to be divisible by this value
    # This is added for computational efficiency reasons.
    make_vocab_size_divisible_by=128,
    data_path=[
        "/workspace/idea_model/idea_bert/output_data/loss_compara_content_sentence"
    ],
    split="949,50,1",
    vocab_file="/workspace/idea_model/idea_bert/bert-base-chinese-vocab.txt",
    ...
)

# bert.py
model = dict(
    model_name="BertModel",
    
  	model_cfg=dict(
        vocab_size=30522,
        hidden_size=768,
        hidden_layers=24,
        num_attention_heads=12,
      ...
    )
)

# gpt.py
model = dict(
		model_name="GPT_2",
  
  	model_cfg=dict(
    		vocab_size=30522,
      	hidden_size=768,
      ...
    )
)

# bert_large.py
from bert.py import model
model.model_cfg.hidden_size = 1537
```

通过这种方式，需要使用哪个模型作为模板可以直接 import 对应的模型配置文件，然后再做对应的修改，如果是一个全新的模型，也可以根据模型本身的参数完全重写一套，保证配置文件中的参数名字和模型定义的参数名字一致就可以了。

要查看内置模型需要的参数也可以直接访问对应模型的配置文件，不需要到模型定义的代码处进行查看。最后序列化会只保存使用到的配置文件参数。

## 模型的配置文件举例

下面分别用上述提到的两种配置方式对 bert 模型构建进行举例说明。

### 采用注册+解析的方式

```python
# configs/model/bert.py

# registry + arguments parse
model = dict(
    model_name="BertModel",
    
  	model_cfg=dict(
        vocab_size=30522,
        hidden_size=768,
        hidden_layers=24,
        num_attention_heads=12,
        intermediate_size=4096,
        hidden_dropout_prob=0.1,
        attention_probs_dropout_prob=0.1,
        max_position_embeddings=512,
        num_tokentypes=2,
        add_pooling_layer=True,
        initializer_range=0.02,
        layernorm_eps=1e-5,
        bias_gelu_fusion=True,
        bias_dropout_fusion=True,
        scale_mask_softmax_fusion=False,
        apply_query_key_layer_scaling=True,
    ),
)

# modify default arguments
model.model_cfg.vocab_size = 3000
```

序列化后的 config.yaml 文件，支持反序列化之后进行构建。

```yaml
# config.yaml
model:
  model_name: BertModel
  model_cfg: {add_pooling_layer: true, apply_query_key_layer_scaling: true, attention_probs_dropout_prob: 0.1, bias_dropout_fusion: true, bias_gelu_fusion: true, hidden_dropout_prob: 0.1, hidden_layers: 24, hidden_size: 768, initializer_range: 0.02, intermediate_size: 4096, layernorm_eps: 1.0e-05, max_position_embeddings: 512, num_attention_heads: 12, num_tokentypes: 2, scale_mask_softmax_fusion: false, vocab_size: 30522}
```

### LazyCall

```python
# configs/model/bert.py

# LazyCall
from libai.config import LazyCall
from libai.models import BertModel

cfg = dict(
    vocab_size=30522,
    hidden_size=768,
    hidden_layers=24,
    num_attention_heads=12,
    intermediate_size=4096,
    hidden_dropout_prob=0.1,
    attention_probs_dropout_prob=0.1,
    max_position_embeddings=512,
    num_tokentypes=2,
    add_pooling_layer=True,
    initializer_range=0.02,
    layernorm_eps=1e-5,
    bias_gelu_fusion=True,
    bias_dropout_fusion=True,
    scale_mask_softmax_fusion=False,
    apply_query_key_layer_scaling=True,
)

model = LazyCall(BertModel)(cfg=cfg)

# modify default arguments
model.cfg.vocab_size = 3000
```

序列化之后的配置文件。

```yaml
model:
  _target_: libai.models.BertModel
  cfg: {add_pooling_layer: true, apply_query_key_layer_scaling: true, attention_probs_dropout_prob: 0.1, bias_dropout_fusion: true, bias_gelu_fusion: true, hidden_dropout_prob: 0.1, hidden_layers: 24, hidden_size: 768, initializer_range: 0.02, intermediate_size: 4096, layernorm_eps: 1.0e-05, max_position_embeddings: 512, num_attention_heads: 12, num_tokentypes: 2, scale_mask_softmax_fusion: false, vocab_size: 30522}
```

上面两种方法最终获得的保存结果是类似的，唯一的区别是 LazyCall 可以更好的定位到使用的模型是 `libai.models.BertModel` 而注册+解析的方式只能获得一个模型名称。

## 总结

有了标准的注册+解析的配置文件机制，最后再讲一下为什么我们还想额外引入 LazyCall 的机制：

1. 使用 LazyCall 的对象构建是与**配置系统无关**的，其只是常见的 python 函数和类，这意味着我们的配置文件甚至可以被其他的库引入，比如

   ```python
   linear1d = {"_target_": "libai.layers.Linear", "in_features": 100, "out_features": 200, "parallel": "col"}
   linear = {"_target_": "flow.nn.Linear", "in_features": 100, "out_features": 200}
   ```

   这个配置文件可以直接在其他的库进行实例化而不依赖于使用 cfg 文件。

2. 通过阅读配置结果，可以更**清晰地**看到调用了哪个类，以及哪个函数，他们分别使用的参数是什么。
3. `cfg` 中的 key 并不需要提前被定义，使用 LazyCall 给了我们机会在对象和函数被实例化之前有机会能修改他的值，这给了我们更多的**灵活性**。

4. LazyCall 和注册+解析是**兼容**的，同时给了我们更多的可能性去探索可编程式的配置文件。

### CODE REVIEW

File: libai/models/bert_model.py [hunk 1]
```diff
@@ -311,6 +313,7 @@ def forward(self, hidden_states, extended_attention_mask):
         return output
 
 
+@MODEL_ARCH_REGISTRY.register()
```
Comment (L1aoXingyu) [inline libai/models/bert_model.py:316]: 对模型进行注册，可以通过名字进行访问

**Comment by strint** [comment]
LazyCall建议改为
- LazyInit
- LazyFactory
因为其功能就是做对象延迟初始化的。

LazyCall感觉和lazy执行有含义的冲突。

**Comment by strint** [comment]
简称L也不是一个可以作为范例的简称，太过简化以至于没有了任何含义。

File: libai/models/build.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+def build_model(cfg):
+    """ Build the whole model architecture, defined by ``cfg.model.model_arch``.
+    Note that is does not load any weights from ``cfg``.
+    """
+    if "_target_" in cfg.model:  # LazyCall
+        model = instantiate(cfg.model)
+    else:
+        model_arch = cfg.model.model_arch
+        model = MODEL_ARCH_REGISTRY.get(model_arch)(cfg.model.model_cfg)
```
Comment (L1aoXingyu) [inline libai/models/build.py:36]: 支持两种兼容的模型构建方式

**Comment by L1aoXingyu** [comment]
## 整体调用流程介绍

1. 首先通过一个完整的 `config.py` 构建配置参数，这个文件可以通过 import 其他已有的文件来减少重复参数的写入，比如

   ```python
   # bert_large_pretrain.py
   
   from .common.train import train # 训练相关的超参，比如 batch_size, train_iter 等参数
   from .common.optim import optim, lr_scheduler # optimizer 和 lr scheduler 构建的超参数
   from .common.data.nlp_data import data # 和数据有关的超参数
   from .common.models.bert_base import model # 和模型相关的超参数
   
   # 针对 bert_large 从 bert_base 修改的超参
   model.cfg.hidden_size = 1536
   model.cfg.hidden_layers = 16
   
   # 修改一些训练超参
   train.micro_batch_size = 32
   train.output_dir = "bert_large_ckpt"
   
   ```

2. 接着通过命令行传参将该配置文件传入，也可以在命令行中修改配置文件中的参数项

   ```shell
   python3 -m oneflow.distributed.launch \
   --nproc_per_node 4 --nnodes 1 --node_rank 0 --master_addr 127.0.0.1 \
   tools/pretrain_bert.py \
   --config-file configs/bert_large_pretrain.py optim.lr=0.01 train.log_period=1
   ```

3. 在主体代码中，会通过 `cfg = LazyConfig.load(args.config_file)` 获得配置文件中的参数，然后通过 `LazyConfig.apply_overrides(cfg, args.opts)` 重载命令行的传参。

4. 在训练之前进行一些基本的设置，主要是下面6个步骤：

   1. 设置 LiBai logger；
   2. 记录一些基本信息，比如配置文件名称，命令行参数，环境变量等等；
   3. 设置分布式环境
      这里会通过 `flow.env.get_node_size(), flow.env.get_world_size()` 来获取实际的 gpu 数量和节点数量，然后和配置文件中对应的参数进行比较
      - 如果是载入自己写的配置文件，默认配置文件这没有这些参数的 key，会被直接增加到配置文件中；
      - 如果是载入别人的配置文件进行训练，这些保存的配置文件会存在这些分布式信息，这是会和这次训练设置的分布式信息进行比较，如果不匹配会给出 warning，提示这种配置和之前的模型训练不一致，可能无法完全复现结果，最后仍然按本次配置的 gpu 和节点进行训练；
   4. 设置 tokenizer 以改变 vocab_size 的大小；
   5. 检查 batch size，设置 gradient accumulation 以及 global batch size 等信息；
   6. 序列化 config 文件，以便复现结果；

**Comment by rentainhe** [comment]
我没有任何意见了，我感觉config system可以尽快合并，因为很多PR都依赖于这个配置系统

File: libai/config/arguments.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    Args:
+        epilog (str): epilog passed to ArgumentParser describing the usage.
+
+    Returns:
+        argparse.ArgumentParser:
+    """
+    parser = argparse.ArgumentParser(
+        epilog=epilog
+        or f"""
+Examples:
```
Comment (dangkai4u) [inline libai/config/arguments.py:34]: 这部分文档，为什么顶格写啊

File: libai/config/config.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+            def __init__(self, a, b=2, c=3):
+                pass
+            @classmethod
+            def from_config(cls, cfg):   # 'cfg' must be the first argument
+                # Returns kwargs to be passed to __init__
+                return {"a": cfg.A, "b": cfg.B}
+        a1 = A(a=1, b=2)  # regular construction
+        a2 = A(cfg)       # construct with a cfg
+        a3 = A(cfg, b=3, c=4)  # construct with extra overwrite
+        # Usage 2: Decorator on any function. Needs an extra from_config argument:
```
Comment (dangkai4u) [inline libai/config/config.py:39]: 这部分适当加一些换行，就像写代码一样，毕竟这也算是教程了

File: libai/models/bert_model.py [hunk 5]
```diff
@@ -334,6 +336,7 @@ def word_embeddings_weight(self):
         return self.embeddings.word_embeddings()
 
 
+@MODEL_ARCH_REGISTRY.register()
```
Comment (dangkai4u) [inline libai/models/bert_model.py:339]: 我们实现的注册机制，不需要加名字吗

File: libai/config/arguments.py (hunk 3 above)
Comment (L1aoXingyu) [reply libai/config/arguments.py:34]: 是为了在 `-h` 下更好得提供下面的信息

image

File: configs/bert_large_pretrain.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
+from libai.config import LazyCall
```
Comment (dangkai4u) [inline configs/bert_large_pretrain.py:1]: 这里没加header

File: libai/models/bert_model.py (hunk 5 above)
Comment (rentainhe) [reply libai/models/bert_model.py:339]: 名字就是定义的那个函数或者class的名字

Comment (L1aoXingyu) [reply libai/models/bert_model.py:339]: 我们的注册机制使用的是 `obj.__name__`

File: configs/bert_large_pretrain.py (hunk 6 above)
Comment (L1aoXingyu) [reply configs/bert_large_pretrain.py:1]: 配置文件都不加 header

File: libai/config/arguments.py (hunk 3 above)
Comment (dangkai4u) [reply libai/config/arguments.py:34]: 为什么是tools目录下的，这些训练的脚本，放在examples文件下，tools打算放一些临时的脚本，例如数据处理、评价指标

Comment (L1aoXingyu) [reply libai/config/arguments.py:34]: examples/projects 下放其他的模型案例，tools 里面放内置模型的训练脚本

Comment (dangkai4u) [reply libai/config/arguments.py:34]: 所有模型的训练脚本或样例应该放在一块，别人的模型如果实现好的话，可以纳入内置的；如果不好，就不合并到该仓库中。

Comment (L1aoXingyu) [reply libai/config/arguments.py:34]: 首先内置模型只需要一个训练的 python 文件，不同的模型只是配置文件不同而已，而其他非内置的模型依赖于他自己的模块，所以不可能用同一个训练脚本，也不能放在一起，你去看 fairseq 的 examples 也是把不同的模型放到在不同的文件夹。

另外我们的内置模型尽量是经典模型，数量少但是经典，而其他的模型都在 project/examples 里面去实现
1. 有一些模型会有特别的 layer，而且只有他会用到，如果要合并到 libai 内部，还需要在 libai 内部额外支持这些 layer，这些 layer 其实并不是常用的，所以 libai 会越来越臃肿；
2. 模型太多了我们维护成本会增加，我们的升级成本也会更高，所有的模型都放在对应的 examples 的文件夹下面，让他们的 owner 去负责，而不是放在 libai 内置让我们去负责；

File: tools/pretrain_bert.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    return tokens, padding_mask, types, sentence_order, lm_labels, loss_mask
+
+
+class Trainer(DefaultTrainer):
+    def run_step(self):
+        return super().run_step(get_batch)
+
+
+def main(args):
+    cfg = LazyConfig.load(args.config_file)
```
Comment (dangkai4u) [inline tools/pretrain_bert.py:59]: 能不能别让用户写这部分代码，格式条款，除了get_batch和用户自定义有关，其他都无关，完全可以固定下来

Comment (dangkai4u) [reply tools/pretrain_bert.py:59]: get_batch函数，是不是可以让用户输入，或者使用参数来区分。数据类型由dataset里决定，key、input_placement、label_placement作为参数指定，最终在dataset里由用户确定

Comment (rentainhe) [reply tools/pretrain_bert.py:59]: 你是指`pretrain_bert.py`这个文件都不需要吗

Comment (dangkai4u) [reply tools/pretrain_bert.py:59]: 创建一个通用的，具体输入是啥，通过参数或dataset属性指定

Comment (rentainhe) [reply tools/pretrain_bert.py:59]: 这很难吧，如果用户要自己定制的话还是需要自己写一个文件的

Comment (dangkai4u) [reply tools/pretrain_bert.py:59]: 学习使用这个trainer，写一大段和用户目标不相关的代码，其实挺难的

**Review by Ldpe2G — APPROVED** [review]
